src/file.coffee

```diff
@@ -215,35 +240,35 @@ class File
   #
   # Returns a promise that resovles to a String.
   read: (flushCache) ->
-    Q ?= require 'q'
-
-    if not @exists()
-      promise = Q(null)
-    else if not @cachedContents? or flushCache
+    if @cachedContents? and not flushCache
+      promise = Q(@cachedContents)
+    else
       deferred = Q.defer()
       promise = deferred.promise
       content = []
       bytesRead = 0
 
-      encoding = @getEncoding()
-      if encoding is 'utf8'
-        readStream = fs.createReadStream(@getPath(), {encoding})
-      else
-        iconv ?= require 'iconv-lite'
-        readStream = fs.createReadStream(@getPath()).pipe(iconv.decodeStream(encoding))
-
-      readStream.on 'data', (chunk) ->
-        content.push(chunk)
-        bytesRead += chunk.length
-        deferred.notify(bytesRead)
-
-      readStream.on 'end', ->
-        deferred.resolve(content.join(''))
-
-      readStream.on 'error', (error) ->
-        deferred.reject(error)
-    else
-      promise = Q(@cachedContents)
+      try
+        encoding = @getEncoding()
+        if encoding is 'utf8'
+          readStream = fs.createReadStream(@getPath(), {encoding})
+        else
+          iconv ?= require 'iconv-lite'
+          readStream = fs.createReadStream(@getPath()).pipe(iconv.decodeStream(encoding))
+
+        readStream.on 'data', (chunk) ->
+          content.push(chunk)
+          bytesRead += chunk.length
+          deferred.notify(bytesRead)
+
+        readStream.on 'end', ->
+          deferred.resolve(content.join(''))
+
+        readStream.on 'error', (error) ->
+          deferred.reject(error)
+      # when the file doesn't exist, fs.createReadStream will error
+      catch
+        promise = Q(null)
```

Comment:
Ah okay, what do you think about then chaining this read with the exists promise?

Maybe pull out the stream code in a helper method to make it a bit more readable.

My concern is how much code is now wrapped in a `try/catch`, if something like `getPath` or the `iconv-lite` require throws a runtime error it will be swallowed and hard to debug.